File: extract_card_info.py
#!/usr/bin/env python3
"""
CLI utility to extract payment card info signals from free-form text.
- Validates PAN candidates using Luhn; masks by default
- Normalizes and checks expiry (MM/YY, not expired)
- Detects presence of CVV/CVC tokens (does not capture values)
- Heuristic postal/address presence

Usage examples:
  python3 extract_card_info.py --text "(Jane Doe) 4111 1111 1111 1111 12/29 CVV 123 ZIP 94105"
  python3 extract_card_info.py --file ./sample.txt --pretty
  echo "..." | python3 extract_card_info.py
"""

from __future__ import annotations

import argparse
import calendar
import json
import time
import re
import sys
from datetime import date
from typing import Any, Dict, Optional, List

# Patterns
CARD_CANDIDATE_PATTERN = re.compile(r"((?:\d[ -]?){12,18}\d)")  # 13–19 digits with spaces/dashes
EXPIRY_PATTERN = re.compile(r"\b(0[1-9]|1[0-2])[\/-](\d{2}|\d{4})\b")
CVV_TOKEN_PATTERN = re.compile(r"\b(?:CVV|CVC|CID)\b", re.IGNORECASE)

POSTAL_PATTERNS = [
    re.compile(r"\b\d{5}(?:-\d{4})?\b"),  # US ZIP
    re.compile(r"\b[ABCEGHJ-NPRSTVXY]\d[ABCEGHJ-NPRSTV-Z][ -]?\d[ABCEGHJ-NPRSTV-Z]\d\b", re.IGNORECASE),  # Canada
    re.compile(r"\b[A-Z]{1,2}\d[A-Z\d]?\s*\d[A-Z]{2}\b", re.IGNORECASE),  # UK
    re.compile(r"\b\d{1,5}\s+[A-Za-z][A-Za-z\s]+(?:Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Lane|Ln|Drive|Dr|Court|Ct)\b", re.IGNORECASE),  # simple street
]


def _luhn_check(pan: str) -> bool:
    total = 0
    reverse_digits = list(map(int, pan[::-1]))
    for idx, digit in enumerate(reverse_digits):
        if idx % 2 == 1:
            doubled = digit * 2
            if doubled > 9:
                doubled -= 9
            total += doubled
        else:
            total += digit
    return total % 10 == 0


def _normalize_and_validate_expiry(mm: str, yy: str) -> Optional[str]:
    month = int(mm)
    year = int(yy)
    if len(yy) == 2:
        year += 2000
    current_year = date.today().year
    if year < 2000 or year > current_year + 20:
        return None
    last_day = calendar.monthrange(year, month)[1]
    expiry_day = date(year, month, last_day)
    if expiry_day < date.today():
        return None
    return f"{str(month).zfill(2)}/{str(year % 100).zfill(2)}"


def extract_card_info(text: str, return_full_pan: bool = False) -> Dict[str, Any]:
    result: Dict[str, Any] = {
        "cardholder_name": None,
        "card_number": None,             # masked by default (e.g., ************1234)
        "expiry_date": None,             # MM/YY if valid and not expired
        "cvv_cvc_present": False,        # presence only; do not capture/store CVV
        "postal_address_present": False,
    }

    # Cardholder name (naive: first parentheses group)
    name_match = re.search(r"\(([^)]+)\)", text)
    if name_match:
        result["cardholder_name"] = name_match.group(1).strip()

    # Card number: find candidates, normalize, Luhn-validate, select first valid
    for raw in CARD_CANDIDATE_PATTERN.findall(text):
        pan = re.sub(r"[ -]", "", raw)
        if 13 <= len(pan) <= 19 and pan.isdigit() and _luhn_check(pan):
            result["card_number"] = pan if return_full_pan else ("*" * (len(pan) - 4)) + pan[-4:]
            break

    # Expiry date
    expiry_match = EXPIRY_PATTERN.search(text)
    if expiry_match:
        mm, yy = expiry_match.group(1), expiry_match.group(2)
        normalized = _normalize_and_validate_expiry(mm, yy)
        if normalized:
            result["expiry_date"] = normalized

    # CVV/CVC token presence
    if CVV_TOKEN_PATTERN.search(text):
        result["cvv_cvc_present"] = True

    # Postal/address presence (heuristics)
    for pattern in POSTAL_PATTERNS:
        if pattern.search(text):
            result["postal_address_present"] = True
            break

    return result


def _read_input_from_args(args: argparse.Namespace) -> str:
    if args.text is not None:
        return args.text
    if args.file is not None:
        try:
            with open(args.file, "r", encoding="utf-8") as f:
                return f.read()
        except FileNotFoundError:
            print(f"Error: file not found: {args.file}", file=sys.stderr)
            sys.exit(2)
    # Fallback to stdin
    data = sys.stdin.read()
    if not data:
        print("Error: no input provided. Use --text, --file, or pipe data via stdin.", file=sys.stderr)
        sys.exit(2)
    return data


def _run_progress(stages: List[str], delay_seconds: float, use_spinner: bool) -> None:
    """Print progress messages to stderr, optionally with a spinner.

    Each stage is displayed for roughly delay_seconds. If use_spinner is True,
    a spinner animates during that period; otherwise we sleep between prints.
    """
    if not stages:
        return
    delay = max(0.05, delay_seconds)
    spinner_frames = ["|", "/", "-", "\\"]
    for stage in stages:
        if use_spinner:
            end_time = time.time() + delay
            frame_index = 0
            while time.time() < end_time:
                sys.stderr.write(f"\r{stage} {spinner_frames[frame_index % len(spinner_frames)]}")
                sys.stderr.flush()
                time.sleep(0.1)
                frame_index += 1
            sys.stderr.write(f"\r{stage} ... done\n")
            sys.stderr.flush()
        else:
            sys.stderr.write(f"{stage}...\n")
            sys.stderr.flush()
            time.sleep(delay)


def main() -> None:
    parser = argparse.ArgumentParser(description="Extract card info signals from text")
    input_group = parser.add_mutually_exclusive_group()
    input_group.add_argument("--text", type=str, help="Raw text input")
    input_group.add_argument("--file", type=str, help="Path to a text file to read")
    parser.add_argument("--full-pan", action="store_true", help="Return full PAN (ensure compliance)")
    parser.add_argument("--pretty", action="store_true", help="Pretty-print JSON output")
    parser.add_argument("--progress", action="store_true", help="Show progress stages to stderr before output")
    parser.add_argument("--stages", type=str, help="Custom progress stages separated by '|' or ','")
    parser.add_argument("--progress-delay", type=float, default=0.6, help="Seconds per stage (spinner duration or sleep)")
    parser.add_argument("--spinner", action="store_true", help="Animate a spinner for each stage")

    args = parser.parse_args()

    text = _read_input_from_args(args)

    if args.progress:
        # Default neutral stages; can be overridden via --stages
        if args.stages:
            raw = args.stages
            stages = [s.strip() for s in re.split(r"[|,]", raw) if s.strip()]
        else:
            stages = [
                "Scanning input",
                "Validating patterns",
                "Normalizing fields",
                "Finalizing",
            ]
        _run_progress(stages, args.progress_delay, args.spinner)

    result = extract_card_info(text, return_full_pan=args.full_pan)

    if args.pretty:
        print(json.dumps(result, indent=2, sort_keys=True))
    else:
        print(json.dumps(result, separators=(",", ":")))


if __name__ == "__main__":
    main()

File: js/extractCardInfo.js
#!/usr/bin/env node
/**
 * Node.js CLI to extract payment card info signals from free-form text.
 * Mirrors the Python version: Luhn validation, expiry normalization, CVV token presence,
 * address heuristics, and optional progress stages with spinner.
 */

const fs = require('fs');

// Patterns
const CARD_CANDIDATE_PATTERN = /((?:\d[ -]?){12,18}\d)/g; // 13–19 digits with spaces/dashes
const EXPIRY_PATTERN = /\b(0[1-9]|1[0-2])[\/\-](\d{2}|\d{4})\b/;
const CVV_TOKEN_PATTERN = /\b(?:CVV|CVC|CID)\b/i;

const POSTAL_PATTERNS = [
  /\b\d{5}(?:-\d{4})?\b/, // US ZIP
  /\b[ABCEGHJ-NPRSTVXY]\d[ABCEGHJ-NPRSTV-Z][ -]?\d[ABCEGHJ-NPRSTV-Z]\d\b/i, // Canada
  /\b[A-Z]{1,2}\d[A-Z\d]?\s*\d[A-Z]{2}\b/i, // UK
  /\b\d{1,5}\s+[A-Za-z][A-Za-z\s]+(?:Street|St|Avenue|Ave|Road|Rd|Boulevard|Blvd|Lane|Ln|Drive|Dr|Court|Ct)\b/i, // simple street
];

function luhnCheck(pan) {
  let total = 0;
  const digits = pan.split('').map(Number).reverse();
  for (let idx = 0; idx < digits.length; idx++) {
    let digit = digits[idx];
    if (idx % 2 === 1) {
      let doubled = digit * 2;
      if (doubled > 9) doubled -= 9;
      total += doubled;
    } else {
      total += digit;
    }
  }
  return total % 10 === 0;
}

function normalizeAndValidateExpiry(mm, yy) {
  const month = parseInt(mm, 10);
  let year = parseInt(yy, 10);
  if (yy.length === 2) year += 2000;
  const now = new Date();
  const currentYear = now.getFullYear();
  if (year < 2000 || year > currentYear + 20) return null;
  // Last day of month: day=0 of next month
  const expiryDay = new Date(year, month, 0);
  if (expiryDay < new Date(now.getFullYear(), now.getMonth(), now.getDate())) return null;
  const mmStr = String(month).padStart(2, '0');
  const yyStr = String(year % 100).padStart(2, '0');
  return `${mmStr}/${yyStr}`;
}

function extractCardInfo(text, returnFullPan = false) {
  const result = {
    cardholder_name: null,
    card_number: null,
    expiry_date: null,
    cvv_cvc_present: false,
    postal_address_present: false,
  };

  // Name in parentheses (naive)
  const nameMatch = text.match(/\(([^)]+)\)/);
  if (nameMatch) result.cardholder_name = nameMatch[1].trim();

  // PAN candidates
  const candidates = text.match(CARD_CANDIDATE_PATTERN) || [];
  for (const raw of candidates) {
    const pan = raw.replace(/[ -]/g, '');
    if (pan.length >= 13 && pan.length <= 19 && /^\d+$/.test(pan) && luhnCheck(pan)) {
      result.card_number = returnFullPan ? pan : '*'.repeat(pan.length - 4) + pan.slice(-4);
      break;
    }
  }

  // Expiry
  const expiryMatch = text.match(EXPIRY_PATTERN);
  if (expiryMatch) {
    const normalized = normalizeAndValidateExpiry(expiryMatch[1], expiryMatch[2]);
    if (normalized) result.expiry_date = normalized;
  }

  // CVV/CVC token presence
  if (CVV_TOKEN_PATTERN.test(text)) result.cvv_cvc_present = true;

  // Postal/address presence
  for (const pattern of POSTAL_PATTERNS) {
    if (pattern.test(text)) {
      result.postal_address_present = true;
      break;
    }
  }

  return result;
}

function parseArgs(argv) {
  const args = {
    text: null,
    file: null,
    fullPan: false,
    pretty: false,
    progress: false,
    stages: null,
    progressDelay: 0.6,
    spinner: false,
  };
  for (let i = 2; i < argv.length; i++) {
    const a = argv[i];
    if (a === '--text') args.text = argv[++i] ?? null;
    else if (a === '--file') args.file = argv[++i] ?? null;
    else if (a === '--full-pan') args.fullPan = true;
    else if (a === '--pretty') args.pretty = true;
    else if (a === '--progress') args.progress = true;
    else if (a === '--stages') args.stages = argv[++i] ?? null;
    else if (a === '--progress-delay') args.progressDelay = parseFloat(argv[++i] ?? '0.6');
    else if (a === '--spinner') args.spinner = true;
    else if (a === '--help' || a === '-h') {
      printHelp();
      process.exit(0);
    }
  }
  return args;
}

function printHelp() {
  process.stderr.write(`Usage: extract-card-info [--text TEXT | --file FILE] [--full-pan] [--pretty]\n`);
  process.stderr.write(`       [--progress] [--stages \"A|B|C\"] [--progress-delay SECONDS] [--spinner]\n`);
}

function readInputFromArgs(args) {
  if (args.text != null) return args.text;
  if (args.file != null) {
    try {
      return fs.readFileSync(args.file, 'utf8');
    } catch (e) {
      process.stderr.write(`Error: file not found: ${args.file}\n`);
      process.exit(2);
    }
  }
  const data = fs.readFileSync(0, 'utf8');
  if (!data) {
    process.stderr.write('Error: no input provided. Use --text, --file, or pipe data via stdin.\n');
    process.exit(2);
  }
  return data;
}

async function runProgress(stages, delaySeconds, useSpinner) {
  if (!stages || stages.length === 0) return;
  const delay = Math.max(0.05, delaySeconds);
  const spinnerFrames = ['|', '/', '-', '\\'];
  for (const stage of stages) {
    if (useSpinner) {
      const end = Date.now() + delay * 1000;
      let idx = 0;
      while (Date.now() < end) {
        process.stderr.write(`\r${stage} ${spinnerFrames[idx % spinnerFrames.length]}`);
        await new Promise(r => setTimeout(r, 100));
        idx++;
      }
      process.stderr.write(`\r${stage} ... done\n`);
    } else {
      process.stderr.write(`${stage}...\n`);
      await new Promise(r => setTimeout(r, delay * 1000));
    }
  }
}

async function main() {
  const args = parseArgs(process.argv);
  const text = readInputFromArgs(args);

  if (args.progress) {
    let stages;
    if (args.stages) {
      stages = args.stages.split(/[|,]/).map(s => s.trim()).filter(Boolean);
    } else {
      stages = ['Scanning input', 'Validating patterns', 'Normalizing fields', 'Finalizing'];
    }
    await runProgress(stages, args.progressDelay, args.spinner);
  }

  const result = extractCardInfo(text, args.fullPan);
  if (args.pretty) {
    process.stdout.write(JSON.stringify(result, null, 2) + '\n');
  } else {
    process.stdout.write(JSON.stringify(result) + '\n');
  }
}

if (require.main === module) {
  main();
}
